Login not possible when using dex with multiple connectors

**Issue Summary**

Hi,
we enabled the openid auth provider and configured it to use dex as openid provider.

Dex is a federated openid provider and we enabled 2 connectors on it(ldap,microsoft)
We could login through dex using the microsoft federated connector, a user is created, and we are able to logout and login again using the same federated connector in dex.

But if we try to login through dex using a different connector(ldap), we get an exception from alerta

```
Oct 15 10:45:04 influx alertad[28913]: 2020-10-15 10:45:04,542 alerta.app[28919]: [ERROR] duplicate key value violates unique constraint "users_email_key"
                                       DETAIL:  Key (email)=(xxxxxx@xxxxx) already exists.
                                        request_id=7f8eab74-9c03-4d36-a92f-c5c0d82b3c6f ip=80.120.197.114
                                       Traceback (most recent call last):
                                         File "/opt/alerta/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
                                           rv = self.dispatch_request()
                                         File "/opt/alerta/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
                                           return self.view_functions[rule.endpoint](**req.view_args)
                                         File "/opt/alerta/lib/python3.7/site-packages/flask_cors/decorator.py", line 128, in wrapped_function
                                           resp = make_response(f(*args, **kwargs))
                                         File "/opt/alerta/lib/python3.7/site-packages/alerta/auth/oidc.py", line 135, in openid
                                           user.create()
                                         File "/opt/alerta/lib/python3.7/site-packages/alerta/models/user.py", line 140, in create
                                           return User.from_db(db.create_user(self))
                                         File "/opt/alerta/lib/python3.7/site-packages/alerta/database/backends/postgres/base.py", line 910, in create_user
                                           return self._insert(insert, vars(user))
                                         File "/opt/alerta/lib/python3.7/site-packages/alerta/database/backends/postgres/base.py", line 1389, in _insert
                                           cursor.execute(query, vars)
                                         File "/opt/alerta/lib/python3.7/site-packages/psycopg2/extras.py", line 327, in execute
                                           return super(NamedTupleCursor, self).execute(query, vars)
                                       psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "users_email_key"
                                       DETAIL:  Key (email)=(xxxxxx@xxxxxxxx) already exists.
Oct 15 10:45:04 influx alertad[28913]: 2020-10-15 10:45:04,541 alerta.app[28919]: [DEBUG] ****************************************

                                                   INSERT INTO users (id, name, login, password, email, status, roles, attributes,
                                                       create_time, last_login, text, update_time, email_verified)
                                                   VALUES ('Cil1aWQ9c2hlZGVuaWdnLG91PVBlb3BsZSxkYz1pbnZlbml1bSxkYz1pbxIEbGRhcA', 'xxxxxxx', 'xxxxxxxxx', '', 'xxxxxxxxxx', 'active', ARRAY['admin']
                                                       NULL, '', '2020-10-15T10:45:04.541Z', true)
                                                   RETURNING *

                                       **************************************** request_id=7f8eab74-9c03-4d36-a92f-c5c0d82b3c6f ip=xxxxx
Oct 15 10:45:04 influx alertad[28913]: 2020-10-15 10:45:04,540 alerta.app[28919]: [DEBUG] ****************************************
                                       SELECT * FROM users WHERE id='Cil1aWQ9c2hlZGVuaWdnLG91PVBlb3BsZSxkYz1pbnZlbml1bSxkYz1pbxIEbGRhcA'
                                       **************************************** request_id=7f8eab74-9c03-4d36-a92f-c5c0d82b3c6f ip=xxxxx

```
The problem is that dex includes the provider when calculating the tokens sub claim, resulting in different sub claims depending on the federated connector used and as alerta uses the sub claim as userid and as search criteria it tries to create a new user with the same email address which produces the error.

I think the solution could be to just use the users email when checking if the user exists in the openid auth module !?

**Environment**
- OS: Linux(Debain 10)
- API version: 8.0.3
- Deployment: self-hosted
- WSGI environment: nginx/uwsgi

- Database: Postgres

- Server config:
 Auth enabled? Yes
 Auth provider? OpenID
 Customer views? No
 
- CLI version: 8.0.0

--- a/alerta/auth/oidc.py
+++ b/alerta/auth/oidc.py
@@ -128,16 +128,23 @@ def openid():
     if not login:
         raise ApiError("Must support one of the following OpenID claims: 'preferred_username', 'nickname' or 'email'", 400)
 
-    user = User.find_by_id(id=subject)
+    if current_app.config['OIDC_LINK_USER_EMAIL'] and email and email_verified:
+        user = User.find_by_email(email=email)
+    else:
+        user = User.find_by_id(id=subject)
+
     if not user:
         user = User(id=subject, name=name, login=login, password='', email=email,
                     roles=current_app.config['USER_ROLES'], text='', email_verified=email_verified)
         user.create()
     else:
-        user.update(login=login, email=email)
+        user.update(login=login, email=email, email_verified=email_verified)
 
     roles = custom_claims[role_claim] + user.roles
     groups = custom_claims[group_claim]
+
+    if user.id != subject:
+        custom_claims['oid'] = user.id  # if subject differs store the original subject as "oid" claim
 
     if user.status != 'active':
         raise ApiError('User {} is not active'.format(login), 403)

--- a/alerta/models/token.py
+++ b/alerta/models/token.py
@@ -37,6 +37,7 @@ class Jwt:
         self.email_verified = kwargs.get('email_verified')
         self.picture = kwargs.get('picture')
         self.customers = kwargs.get('customers')
+        self.oid = kwargs.get('oid')
 
     @classmethod
     def parse(cls, token: str, key: str = None, verify: bool = True, algorithm: str = 'HS256') -> 'Jwt':
@@ -70,7 +71,8 @@ class Jwt:
             scopes=json.get('scope', '').split(' '),  # eg. scope='read write' => scopes=['read', 'write']
             email_verified=json.get('email_verified', None),
             picture=json.get('picture', None),
-            customers=[json['customer']] if 'customer' in json else json.get('customers', list())
+            customers=[json['customer']] if 'customer' in json else json.get('customers', list()),
+            oid=json.get('oid')
         )
 
     @property
@@ -108,6 +110,8 @@ class Jwt:
             data['picture'] = self.picture
         if current_app.config['CUSTOMER_VIEWS']:
             data['customers'] = self.customers
+        if self.oid:
+            data['oid'] = self.oid
         return data
 
     @property

--- a/alerta/settings.py
+++ b/alerta/settings.py
@@ -135,6 +135,7 @@ OIDC_VERIFY_TOKEN = False
 OIDC_ROLE_CLAIM = OIDC_CUSTOM_CLAIM = 'roles'  # JWT claim name whose value is used in role mapping
 OIDC_GROUP_CLAIM = 'groups'  # JWT claim name whose value is used in customer mapping
 ALLOWED_OIDC_ROLES = ALLOWED_GITLAB_GROUPS or ALLOWED_KEYCLOAK_ROLES or ['*']
+OIDC_LINK_USER_EMAIL = True  # if using federated IdP link user accounts by verified email addresses
 
 # SAML 2.0
 SAML2_ENTITY_ID = None

--- a/tests/test_providers.py
+++ b/tests/test_providers.py
@@ -1477,6 +1477,214 @@ class AuthProvidersTestCase(unittest.TestCase):
         self.assertEqual(claims['customers'], ['Foo Corp'], claims)
 
     @requests_mock.mock()
+    def test_openid_idp(self, m):
+
+        test_config = {
+            'TESTING': True,
+            'AUTH_REQUIRED': True,
+            'AUTH_PROVIDER': 'openid',
+            'OIDC_ISSUER_URL': 'https://dev-66191jdr.eu.auth0.com/',
+        }
+
+        authorization_grant = """
+        {
+          "code": "q79ntL3d9mzPFAU-",
+          "clientId": "LMMEZSlPYKvM14cFxnWNWkC4DgvRk0dZ",
+          "redirectUri": "http://local.alerta.io:8000",
+          "state": "051pvr3x5e4y"
+        }
+        """
+
+        discovery_doc = """
+        {
+          "issuer": "https://dev-66191jdr.eu.auth0.com/",
+          "authorization_endpoint": "https://dev-66191jdr.eu.auth0.com/authorize",
+          "token_endpoint": "https://dev-66191jdr.eu.auth0.com/oauth/token",
+          "userinfo_endpoint": "https://dev-66191jdr.eu.auth0.com/userinfo",
+          "mfa_challenge_endpoint": "https://dev-66191jdr.eu.auth0.com/mfa/challenge",
+          "jwks_uri": "https://dev-66191jdr.eu.auth0.com/.well-known/jwks.json",
+          "registration_endpoint": "https://dev-66191jdr.eu.auth0.com/oidc/register",
+          "revocation_endpoint": "https://dev-66191jdr.eu.auth0.com/oauth/revoke",
+          "scopes_supported": [
+            "openid",
+            "profile",
+            "offline_access",
+            "name",
+            "given_name",
+            "family_name",
+            "nickname",
+            "email",
+            "email_verified",
+            "picture",
+            "created_at",
+            "identities",
+            "phone",
+            "address"
+          ],
+          "response_types_supported": [
+            "code",
+            "token",
+            "id_token",
+            "code token",
+            "code id_token",
+            "token id_token",
+            "code token id_token"
+          ],
+          "response_modes_supported": [
+            "query",
+            "fragment",
+            "form_post"
+          ],
+          "subject_types_supported": [
+            "public"
+          ],
+          "id_token_signing_alg_values_supported": [
+            "HS256",
+            "RS256"
+          ],
+          "token_endpoint_auth_methods_supported": [
+            "client_secret_basic",
+            "client_secret_post"
+          ],
+          "claims_supported": [
+            "aud",
+            "auth_time",
+            "created_at",
+            "email",
+            "email_verified",
+            "exp",
+            "family_name",
+            "given_name",
+            "iat",
+            "identities",
+            "iss",
+            "name",
+            "nickname",
+            "phone_number",
+            "picture",
+            "sub"
+          ],
+          "request_uri_parameter_supported": false
+        }
+        """
+
+        jwks_uri = """
+        {
+          "keys": [
+            {
+              "alg": "RS256",
+              "kty": "RSA",
+              "use": "sig",
+              "x5c": [
+                "MIIDDTCCAfWgAwIBAgIJQGmNBJxybcLTMA0GCSqGSIb3DQEBCwUAMCQxIjAgBgNVBAMTGWRldi02NjE5MWpkci5ldS5hdXRoMC5jb20wHhcNMTkwMzMwMDYzODM1WhcNMzIxMjA2MDYzODM1WjAkMSIwIAYDVQQDExlkZXYtNjYxOTFqZHIuZXUuYXV0aDAuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAp2ZcAYOMRWIVqO8Ih3qiLXoL0antgVSsyJvlHmvWJN0Oyn9xkmXX3qbJRIYN13Dj2EW0A1mGZYZBHVPDCoAe7dWLFA8UyU8ofQofQ76AXn4zEhI5ETnj5ZkqSMTfZuA8sbKtLtO7alVTPok24a4sfm4ECG8k5Rx59UOX1JIVOsvQLnWhnxYmoe+fB52XqnhY6tXF3Y+v+QC2jCWo1yZpG/ZpegMu4WcmT3zqxWqT4pbamCUH49IfJbVhRISF9JQdytz/ylfzaow4+QTgpMQHx27L99AFhgL9moxglyMmI8OO8yyoabWdkE9GPtKYiaNwxlfagze6sA0XMuWAMAH3NwIDAQABo0IwQDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBQ4VB9WhbNhas/0DpOGSvQqui8j6zAOBgNVHQ8BAf8EBAMCAoQwDQYJKoZIhvcNAQELBQADggEBACix5lFpyx46oTNwwsJKVo3+gZAlLY6/sHldbUpV8MQIC2HXwCjNQAldUWQkMVWSieU3eE4fEUrfRxIwKOXZ/hV5oA9PSnnVAJyRzXuzYijtKPV15D3InnAr3TsgLGw3kWjPUDgXecqxsIo8hnuYaVIcNPJT12mBlZYtBckf4DAMl76/xwy0Y5vdLZX8/MICl8MoXn4a4DXfhyD/SG0mLYUOepdxAKjZNV5EfTQgUVkiZlYRkiKEU0jRy1moRbXDwJ4r7t1sVASO14D8r8JfkPUhqFovijVmk5+dLbLEQlr6mbEe6221riOD2a/Eq8CTukkYzzXxSDfohVniqHO3mLA="
+              ],
+              "n": "p2ZcAYOMRWIVqO8Ih3qiLXoL0antgVSsyJvlHmvWJN0Oyn9xkmXX3qbJRIYN13Dj2EW0A1mGZYZBHVPDCoAe7dWLFA8UyU8ofQofQ76AXn4zEhI5ETnj5ZkqSMTfZuA8sbKtLtO7alVTPok24a4sfm4ECG8k5Rx59UOX1JIVOsvQLnWhnxYmoe-fB52XqnhY6tXF3Y-v-QC2jCWo1yZpG_ZpegMu4WcmT3zqxWqT4pbamCUH49IfJbVhRISF9JQdytz_ylfzaow4-QTgpMQHx27L99AFhgL9moxglyMmI8OO8yyoabWdkE9GPtKYiaNwxlfagze6sA0XMuWAMAH3Nw",
+              "e": "AQAB",
+              "kid": "MjdFNjIxNjhFN0Y4N0E4RkFCODUyQzAyMUNBQzkxQjNFMDI2MkUyMA",
+              "x5t": "MjdFNjIxNjhFN0Y4N0E4RkFCODUyQzAyMUNBQzkxQjNFMDI2MkUyMA"
+            }
+          ]
+        }
+        """
+
+        # using GitHub credentials
+        access_token = """
+        {
+          "access_token": "2WkWGsYR6dOdFLFAME3LI-21l15BRFiL",
+          "id_token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1qZEZOakl4TmpoRk4wWTROMEU0UmtGQ09EVXlRekF5TVVOQlF6a3hRak5GTURJMk1rVXlNQSJ9.eyJuaWNrbmFtZSI6InNhdHRlcmx5IiwibmFtZSI6Ik5pY2sgU2F0dGVybHkiLCJwaWN0dXJlIjoiaHR0cHM6Ly9hdmF0YXJzMC5naXRodWJ1c2VyY29udGVudC5jb20vdS82MTUwNTc_dj00IiwidXBkYXRlZF9hdCI6IjIwMjAtMTEtMDlUMDA6MDY6MjAuMDgxWiIsImVtYWlsIjoibmZzYXR0ZXJseUBnbWFpbC5jb20iLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwiaXNzIjoiaHR0cHM6Ly9kZXYtNjYxOTFqZHIuZXUuYXV0aDAuY29tLyIsInN1YiI6ImdpdGh1Ynw2MTUwNTciLCJhdWQiOiJMTU1FWlNsUFlLdk0xNGNGeG5XTldrQzREZ3ZSazBkWiIsImlhdCI6MTYwNDg4MDM4MSwiZXhwIjoxNjA0OTE2MzgxfQ.DslYi1zR2w9PYKX25jh9RMQQXVtoOS21opgl0RA0DZG98WjRztzufeSMjDydyYFrQevB739jItWxXfZU_R1uXwvMghsLOdOqT95roWoFWKQW7Z3o35KI3-9XdkLTQ9E5_IGl8U8J666alsF4kns2qUUeV57xoN3BVPFjVr1mM65tO5xWlGD3Mq3sqEVC0OwxtDelG0C4WgGnSmJQryRbaXKo0kl8ixuEDET2aBqBjXjIZ-Av73IkBBpicAsigGlxjDHxl9y6lmkhuH2ft1f1CZO0XBwFFB7VeX2mozRk9lEMudQVUXaAWZQGZPjKt-M9HbH0SwMv4vm6mchG9jkNUA",
+         "scope": "openid profile email",
+         "expires_in": 86400,
+          "token_type": "Bearer"
+        }
+        """
+
+        userinfo = """
+        {
+          "sub": "github|615057",
+          "nickname": "satterly",
+          "name": "Nick Satterly",
+          "picture": "https://avatars0.githubusercontent.com/u/615057?v=4",
+          "updated_at": "2020-11-09T00:06:20.081Z",
+          "email": "nfsatterly@gmail.com",
+          "email_verified": true
+        }
+        """
+
+        m.get('https://dev-66191jdr.eu.auth0.com/.well-known/openid-configuration', text=discovery_doc)
+        m.get('https://dev-66191jdr.eu.auth0.com/.well-known/jwks.json', text=jwks_uri)
+
+        m.post('https://dev-66191jdr.eu.auth0.com/oauth/token', text=access_token)
+        m.get('https://dev-66191jdr.eu.auth0.com/userinfo', text=userinfo)
+
+        self.app = create_app(test_config)
+        self.client = self.app.test_client()
+
+        with self.app.test_request_context('/'):
+            self.app.preprocess_request()
+            self.api_key = ApiKey(
+                user='admin@alerta.io',
+                scopes=[Scope.admin, Scope.read, Scope.write],
+                text='demo-key'
+            )
+            self.api_key.create()
+
+        self.headers = {
+            'Authorization': 'Key %s' % self.api_key.key,
+            'Content-type': 'application/json'
+        }
+
+        response = self.client.post('/auth/openid', data=authorization_grant, content_type='application/json')
+        self.assertEqual(response.status_code, 200, response.data)
+        data = json.loads(response.data.decode('utf-8'))
+        claims = jwt.decode(data['token'], verify=False)
+
+        self.assertEqual(claims['sub'], 'github|615057', claims)
+        self.assertNotIn('oid', claims)
+        self.assertEqual(claims['preferred_username'], 'satterly', claims)
+        self.assertEqual(claims['email'], 'nfsatterly@gmail.com', claims)
+        self.assertEqual(claims.get('email_verified'), True, claims)
+
+        # using Google credentials
+        access_token = """
+        {
+          "access_token": "Ua37VMJg6W-DuQuO14GyDQlidhUXCNzT",
+          "id_token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1qZEZOakl4TmpoRk4wWTROMEU0UmtGQ09EVXlRekF5TVVOQlF6a3hRak5GTURJMk1rVXlNQSJ9.eyJnaXZlbl9uYW1lIjoiTmljayIsImZhbWlseV9uYW1lIjoiU2F0dGVybHkiLCJuaWNrbmFtZSI6Im5mc2F0dGVybHkiLCJuYW1lIjoiTmljayBTYXR0ZXJseSIsInBpY3R1cmUiOiJodHRwczovL2xoMy5nb29nbGV1c2VyY29udGVudC5jb20vYS0vQU9oMTRHaFZrS3pQbXI0RnE4cjRIN3R0eTdCc2g5VWY2WlFXZnFVeFJTOVluUT1zOTYtYyIsImxvY2FsZSI6ImVuLUdCIiwidXBkYXRlZF9hdCI6IjIwMjAtMTEtMDlUMDA6MTI6MjYuODQ4WiIsImVtYWlsIjoibmZzYXR0ZXJseUBnbWFpbC5jb20iLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwiaXNzIjoiaHR0cHM6Ly9kZXYtNjYxOTFqZHIuZXUuYXV0aDAuY29tLyIsInN1YiI6Imdvb2dsZS1vYXV0aDJ8MTA0MDA0NzY0ODI0MDY2MzU5MzkwIiwiYXVkIjoiTE1NRVpTbFBZS3ZNMTRjRnhuV05Xa0M0RGd2UmswZFoiLCJpYXQiOjE2MDQ4ODA3NDcsImV4cCI6MTYwNDkxNjc0N30.IV_SplyKCwhGHis3gt-tCP1wZ9Enqq5o7IzF6HvP345G2SLVZE8Ue2sOZCqLoVAcWF7nMPPCyjujXXETXCMYoWjcgH67TmlPubG4JZtUkuAoTnyytSVPsm6JQpbeqoFo9urCANuAFuDfuQPUNNupX78EzNOgYmQKY66jZl86Gy28c1FRLOFNzowpGyheBm_tABu4wJTToaVoJhmWGGMsW4oHBoRZ3K2S6UEkkszAp-CtZmTNCKYh-vV9GmW5TXgysAGo0sc0TwyqPzJxLvikDhOYeC8f8uCibNnJ4cmaM1igk1YKAZ97Amom5-AmaE10oHTXP2d-_KRaEsdj5weY1g",
+          "scope": "openid profile email",
+          "expires_in": 86400,
+          "token_type": "Bearer"
+        }
+        """
+
+        userinfo = """
+        {
+          "sub": "google-oauth2|104004764824066359390",
+          "given_name": "Nick",
+          "family_name": "Satterly",
+          "nickname": "nfsatterly",
+          "name": "Nick Satterly",
+          "picture": "https://lh3.googleusercontent.com/a-/AOh14GhVkKzPmr4Fq8r4H7tty7Bsh9Uf6ZQWfqUxRS9YnQ=s96-c",
+          "locale": "en-GB",
+          "updated_at": "2020-11-09T00:12:26.848Z",
+          "email": "nfsatterly@gmail.com",
+          "email_verified": true
+        }
+        """
+
+        m.post('https://dev-66191jdr.eu.auth0.com/oauth/token', text=access_token)
+        m.get('https://dev-66191jdr.eu.auth0.com/userinfo', text=userinfo)
+
+        response = self.client.post('/auth/openid', data=authorization_grant, content_type='application/json')
+        self.assertEqual(response.status_code, 200, response.data)
+        data = json.loads(response.data.decode('utf-8'))
+        claims = jwt.decode(data['token'], verify=False)
+
+        self.assertEqual(claims['sub'], 'google-oauth2|104004764824066359390', claims)
+        self.assertEqual(claims['oid'], 'github|615057', claims)
+        self.assertEqual(claims['preferred_username'], 'nfsatterly', claims)
+        self.assertEqual(claims['email'], 'nfsatterly@gmail.com', claims)
+        self.assertEqual(claims.get('email_verified'), True, claims)
+
+    @requests_mock.mock()
     def test_openid_okta(self, m):
 
         test_config = {

